samples/subsys/ipc/openamp/remote/src/main.c

```diff
@@ -0,0 +1,229 @@
+{
+}
+
+static void virtio_notify(struct virtqueue *vq)
+{
+	u32_t dummy_data = 0x00110011; /* Some data must be provided */
+
+	ipm_send(ipm_handle, 0, 0, &dummy_data, sizeof(dummy_data));
+}
+
+struct virtio_dispatch dispatch = {
+	.get_status = virtio_get_status,
+	.get_features = virtio_get_features,
+	.set_features = virtio_set_features,
+	.notify = virtio_notify,
+};
+
+static K_SEM_DEFINE(data_sem, 0, 1);
+static K_SEM_DEFINE(data_rx_sem, 0, 1);
+
+static void platform_ipm_callback(void *context, u32_t id, volatile void *data)
+{
+	k_sem_give(&data_sem);
+}
+
+int endpoint_cb(struct rpmsg_endpoint *ept, void *data,
+		size_t len, u32_t src, void *priv)
+{
+	received_data = *((unsigned int *) data);
+
+	k_sem_give(&data_rx_sem);
+
+	return RPMSG_SUCCESS;
+}
+
+static K_SEM_DEFINE(ept_sem, 0, 1);
+
+struct rpmsg_endpoint my_ept;
+struct rpmsg_endpoint *ep = &my_ept;
+
+static void rpmsg_service_unbind(struct rpmsg_endpoint *ept)
+{
+	(void)ept;
+	rpmsg_destroy_ept(ep);
+}
+
+static unsigned int receive_message(void)
+{
+	while (k_sem_take(&data_rx_sem, K_NO_WAIT) != 0) {
+		int status = k_sem_take(&data_sem, K_FOREVER);
+
+		if (status == 0) {
+			virtqueue_notification(vq[1]);
+		}
+	}
+	return received_data;
+}
+
+static int send_message(unsigned int message)
+{
+	return rpmsg_send(ep, &message, sizeof(message));
+}
+
+void app_task(void *arg1, void *arg2, void *arg3)
+{
+	ARG_UNUSED(arg1);
+	ARG_UNUSED(arg2);
+	ARG_UNUSED(arg3);
+	int status = 0;
+	struct metal_device *device;
+	struct rpmsg_device *rdev;
+
+	printk("\r\nOpenAMP[remote] demo started\r\n");
+
+	struct metal_init_params metal_params = METAL_INIT_DEFAULTS;
+	metal_init(&metal_params);
+
+	status = metal_register_generic_device(&shm_device);
+	if (status != 0) {
+		printk("Couldn't register shared memory device: %d\n", status);
+	}
+
+	status = metal_device_open("generic", SHM_DEVICE_NAME, &device);
+
+	if (status != 0) {
+		printk("metal_device_open failed: %d\n", status);
+	}
+
+	io = metal_device_io_region(device, 0);
+
+	/* setup IPM */
+	ipm_handle = device_get_binding("MAILBOX_0");
+
+	ipm_register_callback(ipm_handle, platform_ipm_callback, NULL);
+
+	ipm_set_enabled(ipm_handle, 1);
+
+	/* setup vdev */
+	vq[0] = virtqueue_allocate(VRING_SIZE);
+	vq[1] = virtqueue_allocate(VRING_SIZE);
+
+	vdev.role = RPMSG_REMOTE;
+	vdev.vrings_num = VRING_COUNT;
+	vdev.func = &dispatch;
+	rvrings[0].io = io;
+	rvrings[0].info.vaddr = (void *)VRING_TX_ADDRESS;
+	rvrings[0].info.num_descs = VRING_SIZE;
+	rvrings[0].info.align = VRING_ALIGNMENT;
+	rvrings[0].vq = vq[0];
+
+	rvrings[1].io = io;
+	rvrings[1].info.vaddr = (void *)VRING_RX_ADDRESS;
+	rvrings[1].info.num_descs = VRING_SIZE;
+	rvrings[1].info.align = VRING_ALIGNMENT;
+	rvrings[1].vq = vq[1];
+
+	vdev.vrings_info = &rvrings[0];
+
+	/* setup rvdev */
+	rpmsg_init_vdev(&rvdev, &vdev, NULL, io, NULL);
+	rdev = rpmsg_virtio_get_rpmsg_device(&rvdev);
+
+	rpmsg_create_ept(ep, rdev, "k", RPMSG_ADDR_ANY, RPMSG_ADDR_ANY,
```

Comment:
We should check the `rpmsg_create_ept` return.

Reply:
done